### PR TITLE
WIP - content routing

### PR DESCRIPTION
**Pull request by indirectlylit**
work-in-progress on content routing.

some high-level issues:

- Vue.js, the vue router, vuex interact strangely together.
  - Currently the behavior looks like:
 image
  - it would be better if it looked like: image
- no handling of loading/error/not-found states
- vuex state tree
  - seems to contain "too much" information about individual page content - even pages not displayed
  - can be in an inconsistent state during content loading - after nav but before API callback
  - Possibly be moved to computed properties inside the individual pages, or maybe have just a single "page state" object which is swapped out with relevant state for the current page only?

some other quick notes:

- sub-routes didn't seem to be working as expected - didn't load component
- initial URL design didn't take into account distinction between viewing a topic node and content node
- channel ID not being used?
- need progress and thumbsrc info
- parents contain IDs only (no names)
- resources don't seem to be setting the synced flag
- api is very slow - due to "preload"?
- Is there an API for learn page content?
- current use of vuex feels awkward
  - pages should clear while loading - not continue to show prior vuex state
  - individual page states are stored even when not active?
  - resources - how to look up info on individual item?
  - currently "watching" the url parameters for changes in ID - but there's a period where that's inconsistent with the rest of the vuex state

### CODE REVIEW

**Comment by indirectlylit** [comment]
Looks like others have similar confusion: https://github.com/vuejs/vuex/issues/11

Right now my inclination is to find an alternative router with no direct hooks to vue, and use this to trigger vuex actions instead of implicitly changing components.

**Comment by indirectlylit** [comment]
@jamalex @rtibbles - I'm going to merge this in even though it's still a WIP to get people something to at least test content renderers etc with. Feel free to leave code review comments though - I'll submit a new PR with follow-up work.

File: .eslintrc.js [hunk 1]
```diff
@@ -5,8 +5,9 @@ module.exports = {
   plugins: [
     'html'
   ],
-  // add your custom rules here
+  // custom rules
   'rules': {
+    'no-param-reassign': ['error', { 'props': false }], // permit vuex state mutations
```
Comment (rtibbles) [inline .eslintrc.js:10]: :+1:

File: kolibri/plugins/learn/assets/src/actions.js [hunk 2]
```diff
@@ -28,6 +29,23 @@ const fetchFullContent = ({ dispatch }, id) => {
...
+  contents.forEach(content => {
+    if (!content.progress) {
+      logging.warn('"progress" was not included in API response');
+      content.progress = 'unstarted'; // eslint-disable-line no-param-reassign
+    }
+    if (!content.thumbnail) {
+      logging.warn('"thumbnail" was not included in API response');
+      content.files.forEach(file => {
+        if (file.thumbnail) {
+          content.thumbnail = file.storage_url; // eslint-disable-line no-param-reassign
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/actions.js:43]: We should probably make this mapping server side too.

File: kolibri/plugins/learn/assets/src/actions.js [hunk 3]
```diff
@@ -28,6 +29,23 @@ const fetchFullContent = ({ dispatch }, id) => {
 const nodeAssignment = (nodes, dispatch) => {
   const topics = nodes.filter((node) => node.kind === 'topic');
   const contents = nodes.filter((node) => node.kind !== 'topic');
+
+  // clean up API response
+  contents.forEach(content => {
+    if (!content.progress) {
+      logging.warn('"progress" was not included in API response');
+      content.progress = 'unstarted'; // eslint-disable-line no-param-reassign
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/actions.js:37]: I would rather handle this with server side annotation and graceful defaults in the rendering code.

File: kolibri/plugins/learn/assets/src/actions.js [hunk 4]
```diff
@@ -38,6 +56,12 @@ const nodeAssignment = (nodes, dispatch) => {
  * @param {String} id - The id of the model to fetch the children of.
  */
 const fetchNodes = ({ dispatch }, id) => {
+  if (id === undefined) {
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/actions.js:59]: Nice.

Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/actions.js:59]: Also wondering if we should put the bootstrapped topics into the ContentNode cache somehow too.

File: kolibri/plugins/learn/assets/src/vue/content-page/index.vue [hunk 5]
```diff
@@ -0,0 +1,99 @@
+<template>
+
+  <div>
+
+    <breadcrumbs
+      v-if="$route.name === 'explore-content'">
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/vue/content-page/index.vue:6]: I hate this. It seems so unVue. Why can't the router pass in parameters into the components as props?!

(Not your fault, just a rant against the design of Vue Router)

File: kolibri/plugins/learn/assets/src/vue/content-page/index.vue [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    </breadcrumbs>
+    <a
```
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/vue/content-page/index.vue:8]: What I do like, though, is the work you have done to unify the two content pages. Much DRY!

**Pull request merged**